### PR TITLE
Content style toggling

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #1007.

The proposal here is to toggle B/I/U styles on the content and not on the attributes of the element. The reason for this is because element styles cannot be changed by the text editor, which would make styles inconsistent between text selection and the complete text element. From the user point of view, toggling Bold on the side panel for the frame, or cmd+B in the text editor are completely interchangeable.

With this tool, toggling styles becomes:

```
const {toggleFlag} = useContentHelper();
const updateContentStyleToggle = useCallback((flag, value) => {
  pushUpdates(({content}) => {content: setFlag(content, flag, value));
});
...
<BoldButton onClick={(value) => updateContentStyleToggle('b', value)}>
<ItalicButton onClick={(value) => updateContentStyleToggle('i', value)}>
...
```

And I propose we solve the "bold" problem this way:

```
const {content, fontWeight} = props;  // e.g. 100, 400, 700, etc.
const {setCustomStyles} = useContentHelper();
const boldWeight = Math.min(Math.max(fontWeight + 100, 700), 900);
if boldWeight > 700) {
  contentMarkup = setCustomStyles(content, 'b', {fontWeight: boldWeight});
}
```

Basically this will work simply like this:
 1. A 100-600 fonts will get bold of weight 700. That's the default anyhow.
 2. A 700 font will get bold of 800, a 800 font will get bold of 900.
 3. A 900 font will always just get 900. I.e. bold toggle has no effect.

### CODE REVIEW

**Comment by obetomuniz** [comment]
I liked this @dvoytenko. I guess this proposal also solves partially #413, WDYT?

Also, what about adding a way to extract font-weights? It will be useful to bind the Font Weight dropdown as multiple or only a specific one. 

```javascript
const extractFontWeightsFromContentSelected = (dom, tagName, defaultFontWeight) => {
  let fontWeights = new Set()
  const nodeList = dom.querySelectorAll(tagName);
   for (let i = nodeList.length - 1; i >= 0; i--) {
     fontWeights.add(
      parseInt(
        nodeList[i].style.fontWeight ? nodeList[i].style.fontWeight : defaultFontWeight
      )
     );
   }
   return [...fontWeights]
}
```

It's proposing to use only a delimited content selected based on a specific tag, but it also can extract from the whole content, we just need to consider the variants I & U too. Thoughts?

**Comment by dvoytenko** [comment]
> I liked this @dvoytenko. I guess this proposal also solves partially #413, WDYT?

It's mostly unrelated to #413, tbh. That one is specifically only about using toggle buttons on selections. All the same issues can now be encountered by simply pressing cmd+B on the selection.

> Also, what about adding a way to extract font-weights? It will be useful to bind the Font Weight dropdown as multiple or only a specific one.

Maybe. We can grow this set of utilities as needed. However, I don't remember the requirement to apply the font weight selector to selections (or font, or anything besides the toggles). I could be wrong. But I'd still first focus on the toggles.

**Comment by barklund** [comment]
> However, I don't remember the requirement to apply the font weight selector to selections (or font, or anything besides the toggles). I could be wrong. But I'd still first focus on the toggles.

[@pbakaus describes exactly that](https://github.com/google/web-stories-wp/pull/1093#discussion_r406600073):

> 8. If you have any kind of text selection and change the weight via dropdown, everything in the selection switches to that weight.

I don't know I he considered it for all the other text manipulations, though. What about text alignment? Line height? Font family or size? That's suddenly a lot of extra complexity!

**Comment by barklund** [comment]
I do however think this is all too complex for us to handle now, so I say we punt all this to post-alpha. It's a completely new system we need here - I do however believe we can utilize draftjs, as it's very extensible as far as I know (FB OSS).

@pbakaus?

**Comment by pbakaus** [comment]
> > However, I don't remember the requirement to apply the font weight selector to selections (or font, or anything besides the toggles). I could be wrong. But I'd still first focus on the toggles.
> 
> [@pbakaus describes exactly that](https://github.com/google/web-stories-wp/pull/1093#discussion_r406600073):
> 
> > 1. If you have any kind of text selection and change the weight via dropdown, everything in the selection switches to that weight.
> 
> I don't know I he considered it for all the other text manipulations, though. What about text alignment? Line height? Font family or size? That's suddenly a lot of extra complexity!

I think this should work for other properties of text, and definitely for italic, strikethrough and the likes. I'm not sure if we should allow multiple line heights for aesthetic reasons though.

**Comment by pbakaus** [comment]
> I do however think this is all too complex for us to handle now, so I say we punt all this to post-alpha. It's a completely new system we need here - I do however believe we can utilize draftjs, as it's very extensible as far as I know (FB OSS).
> 
> @pbakaus?

post-alpha seems fine - which is post Monday :)

**Comment by obetomuniz** [comment]
> I do however believe we can utilize draftjs, as it's very extensible as far as I know (FB OSS).

We can even explore more the [RichUtils.toggleInlineStyle](https://draftjs.org/docs/api-reference-rich-utils/#toggleinlinestyle) from draftjs.

**Comment by dvoytenko** [comment]
> > I do however believe we can utilize draftjs, as it's very extensible as far as I know (FB OSS).
> 
> We can even explore more the [RichUtils.toggleInlineStyle](https://draftjs.org/docs/api-reference-rich-utils/#toggleinlinestyle) from draftjs.

`RichUtils.toggleInlineStyle` could work and we should definitely try. But I think that will follow a very similar style that I'm suggesting here.

**Comment by dvoytenko** [comment]
@pbakaus 

> I think this should work for other properties of text, and definitely for italic, strikethrough and the likes. I'm not sure if we should allow multiple line heights for aesthetic reasons though.

I think the we are likely to have a lot of complicated nuances with all the different font styles, letter spacing, etc. But I think as soon as practical we need address the basic boolean toggles B/I/U. Mostly because they are broken now and it's very noticeable. More complicated styles could take longer.

**Comment by pbakaus** [comment]
> @pbakaus
> 
> > I think this should work for other properties of text, and definitely for italic, strikethrough and the likes. I'm not sure if we should allow multiple line heights for aesthetic reasons though.
> 
> I think the we are likely to have a lot of complicated nuances with all the different font styles, letter spacing, etc. But I think as soon as practical we need address the basic boolean toggles B/I/U. Mostly because they are broken now and it's very noticeable. More complicated styles could take longer.

Agree.
